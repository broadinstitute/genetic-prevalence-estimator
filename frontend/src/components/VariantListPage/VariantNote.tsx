import { AddIcon, EditIcon, ViewIcon } from "@chakra-ui/icons";
import {
  Button,
  ButtonProps,
  FormControl,
  FormLabel,
  HStack,
  IconButton,
  Modal,
  ModalBody,
  ModalCloseButton,
  ModalContent,
  ModalHeader,
  ModalOverlay,
  Textarea,
  Tooltip,
  VStack,
  useDisclosure,
} from "@chakra-ui/react";
import { useState } from "react";

import { VariantId } from "../../types";

interface VariantNoteFormProps {
  initialNote: string;
  onCancel: () => void;
  onSubmit: (note: string) => void;
}

const VariantNoteForm = (props: VariantNoteFormProps) => {
  const { initialNote, onCancel, onSubmit } = props;
  const [note, setNote] = useState(initialNote);

  return (
    <form
      onSubmit={(e) => {
        e.preventDefault();
        onSubmit(note);
      }}
    >
      <VStack mb={4} spacing={4} align="flex-start">
        <FormControl id="edit-variant-note-note">
          <FormLabel>Note</FormLabel>
          <Textarea
            value={note}
            onChange={(e) => {
              setNote(e.target.value);
            }}
          />
        </FormControl>

        <HStack justify="flex-end" width="100%">
          <Button onClick={onCancel}>Cancel</Button>
          <Button colorScheme="blue" type="submit">
            Submit
          </Button>
        </HStack>
      </VStack>
    </form>
  );
};

type VariantNoteProps = Omit<ButtonProps, "onClick"> & {
  variantId: VariantId;
  note: string | undefined;
  onEdit: (note: string) => void;
  userCanEdit: boolean;
};

export const VariantNote = (props: VariantNoteProps) => {
  const { variantId, note, onEdit, userCanEdit } = props;
  const hasNote = !!note;

  const { isOpen, onOpen, onClose } = useDisclosure();

  return (
    <>
      {!userCanEdit && (
        <>
          {hasNote && (
            <>
              <Tooltip
                hasArrow
                label={
                  <>
                    Note: <p>{note}</p>
                  </>
                }
              >
                <IconButton
                  aria-label={"View note"}
                  icon={<ViewIcon />}
                  size="sm"
                />
              </Tooltip>
            </>
          )}
        </>
      )}
      {userCanEdit && (
        <>
          <Tooltip
            hasArrow
            label={
              hasNote ? (
                <>
                  Edit note<p>{note}</p>
                </>
              ) : (
                "Add note"
              )
            }
          >
            <IconButton
              aria-label={hasNote ? "Edit note" : "Add note"}
              icon={hasNote ? <EditIcon /> : <AddIcon />}
              size="sm"
              onClick={onOpen}
            />
          </Tooltip>
          <Modal isOpen={isOpen} size="xl" onClose={onClose}>
            <ModalOverlay />
            <ModalContent>
              <ModalHeader>{variantId}</ModalHeader>
              <ModalCloseButton />
              <ModalBody>
                <VariantNoteForm
                  initialNote={note || ""}
                  onCancel={onClose}
                  onSubmit={(value) => {
                    onClose();
                    onEdit(value);
                  }}
                />
              </ModalBody>
            </ModalContent>
          </Modal>
        </>
      )}
    </>
  );
};
