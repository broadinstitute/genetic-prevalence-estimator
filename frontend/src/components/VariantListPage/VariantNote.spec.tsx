import { render } from "@testing-library/react";

import { VariantNote } from "./VariantNote";

describe("VariantNote", () => {
  test("is a plus icon button icon if the user has edit permissions and there is no note", () => {
    const result = render(
      <VariantNote
        variantId={"testid123"}
        note={undefined}
        onEdit={() => {
          return;
        }}
        userCanEdit={true}
      />
    );
    expect(result.asFragment()).toMatchSnapshot();
  });

  test("is a pencil icon button icon if the user has edit permissions and there is a note", () => {
    const result = render(
      <VariantNote
        variantId={"testid234"}
        note={"hello world"}
        onEdit={() => {
          return;
        }}
        userCanEdit={true}
      />
    );
    expect(result.asFragment()).toMatchSnapshot();
  });

  test("is blank if the user does not have edit permissions and there is no note", () => {
    const result = render(
      <VariantNote
        variantId={"testid345"}
        note={undefined}
        onEdit={() => {
          return;
        }}
        userCanEdit={false}
      />
    );
    expect(result.asFragment()).toMatchSnapshot();
  });

  test("is an eyes icon if the user does not have edit permissions and there is a note", () => {
    const result = render(
      <VariantNote
        variantId={"testid456"}
        note={"hello world"}
        onEdit={() => {
          return;
        }}
        userCanEdit={false}
      />
    );
    expect(result.asFragment()).toMatchSnapshot();
  });
});
