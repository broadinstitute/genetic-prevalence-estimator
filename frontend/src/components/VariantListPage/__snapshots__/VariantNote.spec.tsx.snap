// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VariantNote is a pencil icon button icon if the user has edit permissions and there is a note 1`] = `
<DocumentFragment>
  <button
    aria-label="Edit note"
    class="chakra-button css-1jyk1h1"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="chakra-icon css-onkibi"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <g
        fill="none"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-width="2"
      >
        <path
          d="M11 4H4a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2v-7"
        />
        <path
          d="M18.5 2.5a2.121 2.121 0 0 1 3 3L12 15l-4 1 1-4 9.5-9.5z"
        />
      </g>
    </svg>
  </button>
</DocumentFragment>
`;

exports[`VariantNote is a plus icon button icon if the user has edit permissions and there is no note 1`] = `
<DocumentFragment>
  <button
    aria-label="Add note"
    class="chakra-button css-1jyk1h1"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="chakra-icon css-onkibi"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M0,12a1.5,1.5,0,0,0,1.5,1.5h8.75a.25.25,0,0,1,.25.25V22.5a1.5,1.5,0,0,0,3,0V13.75a.25.25,0,0,1,.25-.25H22.5a1.5,1.5,0,0,0,0-3H13.75a.25.25,0,0,1-.25-.25V1.5a1.5,1.5,0,0,0-3,0v8.75a.25.25,0,0,1-.25.25H1.5A1.5,1.5,0,0,0,0,12Z"
        fill="currentColor"
      />
    </svg>
  </button>
</DocumentFragment>
`;

exports[`VariantNote is an eyes icon if the user does not have edit permissions and there is a note 1`] = `
<DocumentFragment>
  <button
    aria-label="View note"
    class="chakra-button css-1jyk1h1"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="chakra-icon css-onkibi"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <g
        fill="currentColor"
      >
        <path
          d="M23.432,10.524C20.787,7.614,16.4,4.538,12,4.6,7.6,4.537,3.213,7.615.568,10.524a2.211,2.211,0,0,0,0,2.948C3.182,16.351,7.507,19.4,11.839,19.4h.308c4.347,0,8.671-3.049,11.288-5.929A2.21,2.21,0,0,0,23.432,10.524ZM7.4,12A4.6,4.6,0,1,1,12,16.6,4.6,4.6,0,0,1,7.4,12Z"
        />
        <circle
          cx="12"
          cy="12"
          r="2"
        />
      </g>
    </svg>
  </button>
</DocumentFragment>
`;

exports[`VariantNote is blank if the user does not have edit permissions and there is no note 1`] = `<DocumentFragment />`;
